ntoskrnl/io/iomgr/deviface.c

```diff
@@ -26,192 +27,542 @@ C_ASSERT(sizeof(L"{01234567-89ab-cdef-0123-456789abcdef}") == GUID_STRING_BYTES
+    ReferenceString.Buffer[0] = L'#';
+
+    /* Try to open or create instance subkeys */
+    if (Create)
+    {
+        Status = IopCreateRegistryKeyEx(&InstanceKeyHandle,
+                                        DeviceKeyHandle,
+                                        &ReferenceString,
+                                        DesiredAccess,
+                                        REG_OPTION_NON_VOLATILE,
+                                        &InstanceKeyDisposition);
+    }
+    else
+    {
+        Status = IopOpenRegistryKeyEx(&InstanceKeyHandle,
+                                      DeviceKeyHandle,
+                                      &ReferenceString,
+                                      DesiredAccess);
     }
 
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &ReferenceString,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               *DeviceKeyRealP,
-                               NULL);
-    Status = ZwCreateKey(InstanceKeyRealP,
-                         DesiredAccess,
-                         &ObjectAttributes,
-                         0,
-                         NULL,
-                         REG_OPTION_NON_VOLATILE,
-                         NULL);
     if (!NT_SUCCESS(Status))
     {
-        DPRINT1("Failed to open %wZ%wZ\\%wZ%\\%wZ (%x)\n", &BaseKeyU, &GuidString, &SubKeyName, &ReferenceString, Status);
-        goto cleanup;
+        DPRINT1("Failed to create or open %wZ, Status 0x%08lx\n", &ReferenceString, Status);
+        goto Quit;
     }
 
     Status = STATUS_SUCCESS;
 
-cleanup:
-    if (SubKeyName.Buffer != NULL)
-        ExFreePool(SubKeyName.Buffer);
-
+Quit:
     if (NT_SUCCESS(Status))
     {
-        if (!GuidKey)
-            ZwClose(*GuidKeyRealP);
+        if (DeviceHandle)
+            *DeviceHandle = DeviceKeyHandle;
+        else
+            ZwClose(DeviceKeyHandle);
+
+        if (DeviceDisposition)
+            *DeviceDisposition = DeviceKeyDisposition;
 
-        if (!DeviceKey)
-            ZwClose(*DeviceKeyRealP);
+        if (InstanceHandle)
+            *InstanceHandle = InstanceKeyHandle;
+        else
+            ZwClose(InstanceKeyHandle);
 
-        if (!InstanceKey)
-            ZwClose(*InstanceKeyRealP);
+        if (InstanceDisposition)
+            *InstanceDisposition = InstanceKeyDisposition;
     }
     else
     {
-        if (*GuidKeyRealP != NULL)
-            ZwClose(*GuidKeyRealP);
+        if (InstanceKeyHandle)
+            ZwClose(InstanceKeyHandle);
 
-        if (*DeviceKeyRealP != NULL)
-            ZwClose(*DeviceKeyRealP);
+        if (Create)
+            ZwDeleteKey(DeviceKeyHandle);
 
-        if (*InstanceKeyRealP != NULL)
-            ZwClose(*InstanceKeyRealP);
+        if (DeviceKeyHandle)
+            ZwClose(DeviceKeyHandle);
+
+        if (SymbolicLink.Buffer)
+            ExFreePoolWithTag(SymbolicLink.Buffer, TAG_IO);
```

Comment:
```suggestion
            RtlFreeUnicodeString(&SymbolicLink);
```

Reply:
Shouldn't this string always be freed, both in NT_SUCCESS case and the other?

Reply:
Indeed. Seems it should. But I did not try to free it on success too yet, because I thought it still might be used by something when the function returns. But now I see my guessing appears to be wrong.

Reply:
Done.